aarch64/src/devcons.rs

```diff
@@ -0,0 +1,252 @@
+// Racy to start.
+
+use port::devcons::{Console, Uart};
+
+use core::mem;
+use core::ptr;
+use core::ptr::{read_volatile, write_volatile};
+
+// The aarch64 devcons implementation is focussed on Raspberry Pi 3, 4 for now.
+
+// Useful links
+// - Raspberry Pi Processors
+//     https://www.raspberrypi.com/documentation/computers/processors.html
+// - Raspberry Pi Hardware
+//     https://www.raspberrypi.com/documentation/computers/raspberry-pi.html
+// - Raspi3 BCM2837
+//     Datasheet (BCM2835) https://datasheets.raspberrypi.com/bcm2835/bcm2835-peripherals.pdf
+// - Raspi4 BCM2711
+//     Datasheet https://datasheets.raspberrypi.com/bcm2711/bcm2711-peripherals.pdf
+// - Mailbox
+//     https://github.com/raspberrypi/firmware/wiki/Mailbox-property-interface
+
+// Raspberry Pi has 4 UARTs:
+// - UART0 PL011
+// - UART1 miniUART
+// - UART2 PL011
+// - UART3 PL011
+// We'll be using the UART0.
+
+// TODO
+// - Detect board type and set MMIO base address accordingly
+//     https://wiki.osdev.org/Detecting_Raspberry_Pi_Board
+// - Break out mailbox, gpio code
+
+const MMIO_BASE_BCM2837: u32 = 0x3f000000;
+//const MMIO_BASE_BCM2711: u32 = 0xfe000000;
+
+const GPIO_BASE: u32 = 0x200000;
+const GPPUD: u32 = GPIO_BASE + 0x94;
+const GPPUDCLK0: u32 = GPIO_BASE + 0x98;
+
+// PL011 UART 0 Registers
+const UART0_BASE: u32 = GPIO_BASE + 0x1000;
+const UART0_DR: u32 = UART0_BASE + 0x00; // Data register
+const UART0_FR: u32 = UART0_BASE + 0x18; // Flag register
+const UART0_IBRD: u32 = UART0_BASE + 0x24; // Integer baud rate divisor
+const UART0_FBRD: u32 = UART0_BASE + 0x28; // Fractional baud rate divisor
+const UART0_LCRH: u32 = UART0_BASE + 0x2c; // Line control register
+const UART0_CR: u32 = UART0_BASE + 0x30; // Control register
+const UART0_IMSC: u32 = UART0_BASE + 0x38; // Interrupt mask set clear register
+const UART0_ICR: u32 = UART0_BASE + 0x44; // Interrupt clear register
+
+const MBOX_BASE: u32 = 0xb880;
+const MBOX_READ: u32 = MBOX_BASE + 0x00;
+const MBOX_STATUS: u32 = MBOX_BASE + 0x18;
+const MBOX_WRITE: u32 = MBOX_BASE + 0x20;
+
+const MBOX_FULL: u32 = 0x80000000;
+const MBOX_EMPTY: u32 = 0x40000000;
+
+// Delay for count cycles
+fn delay(count: u32) {
+    unsafe {
+        core::arch::asm!(
+            "1:",
+            "nop",
+            "subs {count}, {count}, 1",
```

Comment:
I saw: `error: invalid instruction mnemonic 'subs'` in the build errors... this is more a question than a nit, is the compiler complaining that it doesn't know about the condition-flag setting operation or is it saying that it wasn't provided an aarch64 "version" to reference known commands from?

Reply:
I was reading the [v8 reference](https://developer.arm.com/documentation/dui0802/a/A64-General-Instructions/SUBS--immediate-)

Reply:
I suspect it was trying to build with an x86-64 toolchain